### ISSUE TITLE
Force System.Guid to map to BINARY(16) by default

### ISSUE THREAD

**Issue by m-wilczynski** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#159]
I would like to suggest to make `System.Guid` field/properties map themselves on migration basis to `BINARY(16)` column types instead of `CHAR(36)`.
Would it be possible to do this by default?
I've tested that mapping using `ModelBuilder` with `.HasColumnType("varbinary(16)")` on property would both produce proper column type for migration and later, yield proper mapping of GUIDs to database and back with no additional actions needed.

Besides the fact that it just doesn't feel right to store GUID/UUID as string, it mostly hurts general DB performance for any Seek action that would use PK or IX.

**Comment by caleblloyd** [comment]
I think it's a good idea. I don't think it should be too hard to change.

I'll have to test behavior with current implementations to see if this will cause a migration of everyone's current `char(36)` to binary. If it does not, then it should be an easy change to put through and publish a release note about.

**Comment by mguinness** [comment]
Check out [connection string options](http://dev.mysql.com/doc/connector-net/en/connector-net-connection-options.html) for Old Guids setting.  It appears that the default changed from `BINARY(16)` to `CHAR(36)`.

**Comment by Kagamine** [comment]
So I vote for char(36)

**Comment by m-wilczynski** [comment]
Shall we compromise then on some sort of setting on configuration basis?
Something like `.WithOldGuids(bool useOldGuids)` that would modify connection string and add `'Old Guids=true'`.
Or is it something that should be just documented and left out for developers to tinker with?

**Comment by Kagamine** [comment]
It was hard to implement because we could not inject connection object in typemapper.

**Comment by m-wilczynski** [comment]
Could you perhaps elaborate on the difficulties you have?
It seems that wrapping around `MySqlConnectionStringBuilder` for concating additional options or just adding some sort of helpers for adding connection string options sounds trival.

Something like:

 ```
services.AddDbContext<AppDb>(
    options => options.UseMySql("SomeConnectionString).WithOldGuids(true));
```

Only problem would be a validation that you are indeed appending to MySQL connection string.

**Comment by Kagamine** [comment]
@caleblloyd Maybe we could make something in UseMySql method, we could read the conn str and add TreatTinyAsInt or others into db options, in this way we could get the options in typemapper. But the datetime(6) issue has not been solved in this way.

**Comment by caleblloyd** [comment]
Can we add an overloaded UseMySql that accepts an Options Enumeration?  So it would look something like:

    options.UseMySql("SomeConnectionString",
        MySqlEntityFrameworkOptions.BinaryGuids
        & MySqlEntityFrameworkOptions.TreatTinyIntAsBool
        & MySqlEntityFrameworkOptions.NoDateTimeDecimal)

Then we can ensure to override the connection string settings accordingly.  If the `MySqlEntityFrameworkOptions.BinaryGuids` option was set, we would ensure `OldGuids=true` in the connection string and if `MySqlEntityFrameworkOptions.TreatTinyIntAsBool` was set, we would ensure that `TreatTinyIntAsBool=true` in the connection string.

I don't think writing a connection string parser is in our interests.  Users could copy and paste a connection string from the internet and not understand the consequences.  I think that we should set our own defaults and then allow for override with an options enumeration.

> Does MySqlConnector support UseOldGuid

Yes it does

**Comment by Kagamine** [comment]
There is no need to add additonal argument, We can get these from the conn str.

**Comment by caleblloyd** [comment]
Sounds good, turns out we already do actually parse the Connection String into a Connection String Builder in [UseMySql()](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/3b600717db2710eec0c3da3c316d5e8c3971fe13/src/Pomelo.EntityFrameworkCore.MySql/Extensions/MySqlDbContextOptionsExtensions.cs#L25-L29)

Think that we could expose the Connection String Builder to be used in DI in other parts of the code, specifically the TypeMapper?

**Comment by caleblloyd** [comment]
Moving to #163 based off design discussion here.

**Comment by Kagamine** [comment]
@m-wilczynski Shipped in 1.1.1-prerelease-10008.
